File: 2019/LevelUp_1May20/lecture001_Recursion/l002.java
import java.util.Arrays;
import java.util.concurrent.TimeUnit;

public class l002{
     public static void main(String[] args){
     solve();
     }


    public static int coinChangePermutation_INF(int[] arr,int lidx, int tar, String ans){
        if(lidx==arr.length || tar==0){
            if(tar==0){
                System.out.println(ans);
                return 1;
            }
            return 0;
        }
    
         int count=0;
         if(tar-arr[lidx]>=0)
             count+=coinChangePermutation_INF(arr,0,tar-arr[lidx],ans+ arr[lidx]+" ");
         count+=coinChangePermutation_INF(arr,lidx+1,tar,ans);
         
         return count;
    }

    
    public static int coinChangePermutation(int[] arr,int lidx, int tar, String ans){
        if(lidx==arr.length || tar==0){
            if(tar==0){
                System.out.println(ans);
                return 1;
            }
            return 0;
        }
    
         int count=0;
        if (arr[lidx] > 0 && tar - arr[lidx] >= 0 ){
            int temp = arr[lidx];
            arr[lidx] = -arr[lidx];
            count+=coinChangePermutation(arr,0,tar-temp,ans + temp+" ");
            arr[lidx] = -arr[lidx];
        }
            count+=coinChangePermutation(arr,lidx+1,tar,ans);
         
         return count;
    }

    
    public static int coinChangeCombination_INF(int[] arr, int lidx, int tar, String ans){
    if(lidx==arr.length || tar==0){
        if(tar==0){
            System.out.println(ans);
            return 1;
        }
        return 0;
    }

     int count=0;
     if(tar-arr[lidx]>=0)
         count+=coinChangeCombination_INF(arr,lidx,tar-arr[lidx],ans+ arr[lidx]+" ");
     count+=coinChangeCombination_INF(arr,lidx+1,tar,ans);
     
     return count;
    }


    public static int coinChangeCombination(int[] arr, int lidx, int tar, String ans){
       if(lidx==arr.length || tar==0){
           if(tar==0){
               System.out.println(ans);
               return 1;
           }
           return 0;
       }

        int count=0;
        if(tar-arr[lidx]>=0)
            count+=coinChangeCombination(arr,lidx+1,tar-arr[lidx],ans+ arr[lidx]+" ");
        count+=coinChangeCombination(arr,lidx+1,tar,ans);
        
        return count;
    }

   //Queens.====================================================================

public static int queensCombination2D(boolean[][] rooms, int room, int tnq, String ans) // qpsf: queen place so far.
{
    if (tnq==0)
    {
        System.out.println(ans);
        return 1;
    }

    int count = 0;
    for (int r = room; r < rooms.length*rooms[0].length; r++){
        int x = r / rooms[0].length;
        int y = r % rooms[0].length;
        count += queensCombination2D(rooms, r + 1,tnq-1, ans + "(" + x + ", " + y + ") ");
    }
    return count;
}

public static int queensPermutation2D(boolean[][] rooms,int tnq, String ans) // qpsf: queen place so far.
{
    if (tnq==0)
    {
        System.out.println(ans);
        return 1;
    }

    int count = 0;
    for (int r = 0; r < rooms.length*rooms[0].length; r++){
        int x = r / rooms[0].length;
        int y = r % rooms[0].length;

        if (!rooms[x][y])
        {
            rooms[x][y] = true;
            count += queensPermutation2D(rooms,tnq-1,  ans + "(" + x + ", " + y + ") ");
            rooms[x][y] = false;
        }
    }
    return count;
}

//NqueenProblem.==================================================

public static boolean isAValidMove(boolean[][] board,int r,int c){
    int[][] dirA={{0,-1},{-1,-1},{-1,0},{-1,1}};
    // int[][] dirA={{0,-1},{-1,-1},{-1,0},{-1,1},{0,1},{1,1},{1,0},{1,-1}};

    for(int d=0;d<dirA.length;d++){
        for(int rad=1;rad<=board.length;rad++){
            int x= r + rad*dirA[d][0];
            int y= c + rad*dirA[d][1];
            
            if(x >= 0 && y >= 0 && x < board.length && y < board[0].length){
                if(board[x][y]) return false;
            }else break;
        }
    }

    return true;
}

public static int Nqueen_01(boolean[][] board, int lpsf, int tnq, String ans) // qpsf: queen place so far.
{
    if (tnq==0)
    {
        System.out.println(ans);
        return 1;
    }

    int count = 0;
    for (int r = lpsf; r < board.length*board[0].length; r++){
        int x = r / board[0].length;
        int y = r % board[0].length;
        if(isAValidMove(board,x,y)){
            board[x][y]=true;
            count += Nqueen_01(board, r + 1,tnq-1, ans + "(" + x + ", " + y + ") ");
            board[x][y]=false;
        }
    }
    return count;
}

public static int Nqueen_02(boolean[][] board, int idx, int tnq, String ans) // qpsf: queen place so far.
{
    if (tnq==0 || idx==board.length*board[0].length)
    {
        if(tnq==0){
           System.out.println(ans);
           return 1;
        }
        return 0;
    }

    int count = 0;
    
    int x = idx / board[0].length;
    int y = idx % board[0].length;
    if(isAValidMove(board,x,y)){
        board[x][y]=true;
        count += Nqueen_02(board, idx + 1,tnq-1, ans + "(" + x + ", " + y + ") ");
        board[x][y]=false;
    }
    count += Nqueen_02(board, idx + 1,tnq, ans);
    return count;
}

public static int Nqueen_03(boolean[][] board, int idx, int tnq, String ans) // qpsf: queen place so far.
{
    if (tnq==0 || idx==board.length*board[0].length)
    {
        if(tnq==0){
           System.out.println(ans);
           return 1;
        }
        return 0;
    }

    int count = 0;
    
    int x = idx / board[0].length;
    int y = idx % board[0].length;
    if(!board[x][y] && isAValidMove(board,x,y)){
        board[x][y]=true;
        count += Nqueen_03(board, 0,tnq-1, ans + "(" + x + ", " + y + ") ");
        board[x][y]=false;
    }
    count += Nqueen_03(board, idx + 1,tnq, ans);
    return count;
}

public static boolean Nqueen_04(boolean[][] rooms,int idx,int tnq, String ans) // qpsf: queen place so far.
{
    if (tnq==0)
    {
        System.out.println(ans);
        return true;
    }

    int count = 0;
    boolean res=false;
    for (int r = idx; r < rooms.length*rooms[0].length; r++){
        int x = r / rooms[0].length;
        int y = r % rooms[0].length;

        if (!rooms[x][y] && isAValidMove(rooms,x,y))
        {
            rooms[x][y] = true;
            res=res || Nqueen_04(rooms,0,tnq-1,  ans + "(" + x + ", " + y + ") ");
            rooms[x][y] = false;
        }
    }
    return res;
}


public static boolean nKnight(int[][] board,int r,int c,int move){
    board[r][c]=move;
    if(move==63){ 
       for(int[] b: board){
        for(int ele: b) System.out.print(ele+ " ");
        System.out.println();
       }
       return true;
   }

   int[] xMove = {  2, 1, -1, -2, -2, -1,  1,  2 }; 
   int[] yMove = {  1, 2,  2,  1, -1, -2, -2, -1 };
   boolean res=false;

   for(int d=0;d<8;d++){
       int x = r + xMove[d];
       int y = c + yMove[d]; 

       if(x >= 0 && y >= 0 && x < 8 && y < 8 && board[x][y]==-1)
           res = res || nKnight(board , x , y , move+1);
    }

    board[r][c]=-1;

    return res;
  
}


//optimised========================================================================

static boolean[] ROW;
static boolean[] COL;
static boolean[] DIAG;
static boolean[] ADIAG;

public static int Nqueen_05(int n,int m,int idx,int tnq, String ans) // qpsf: queen place so far.
{
    if (tnq==0)
    {
        // System.out.println(ans);
        return 1;
    }

    int count = 0;
    for (int r = idx; r < n*m; r++){
        int x = r / m;
        int y = r % m;

        if (!ROW[x] && !COL[y] && !DIAG[x+y] && !ADIAG[x-y+m-1])
        {
            ROW[x]=true; COL[y]=true; DIAG[x+y]=true; ADIAG[x-y+m-1]=true;
            count+= Nqueen_05(n,m,r+1,tnq-1,  ans);
            ROW[x]=false; COL[y]=false; DIAG[x+y]=false; ADIAG[x-y+m-1]=false;           
        }
    }
    return count;
}

static int row=0;
static int col=0;
static int diag=0;
static int adiag=0;

public static int Nqueen_06(int n,int m,int idx,int tnq, String ans) // qpsf: queen place so far.
{
    if (tnq==0)
    {
        // System.out.println(ans);
        return 1;
    }

    int count = 0;
    for (int r = idx; r < n*m; r++){
        int x = r / m;
        int y = r % m;

        if ((row & (1<<x))==0 && (col & (1<<y))==0 && (diag & (1<<(x+y)))==0 && (adiag & (1<<(x-y + m - 1)))==0)
        {
            row^=(1<<x);
            col^=(1<<y);
            diag^=(1<<(x+y));
            adiag^=(1<<(x-y+m-1));

            count+= Nqueen_06(n,m,r+1,tnq-1,  ans );
           
            row^=(1<<x);
            col^=(1<<y);
            diag^=(1<<(x+y));
            adiag^=(1<<(x-y+m-1));
        }
    }
    return count;
}



// ==========================================================================

public static void coinChange(){
    int[] arr={2,3,5,7};
    int tar=10;


   //  System.out.println(coinChangePermutation_INF(arr,0,tar,""));
   //  System.out.println(coinChangePermutation(arr,0,tar,""));
   //  System.out.println(coinChangeCombination_INF(arr,0,tar,""));
    System.out.println(coinChangeCombination(arr,0,tar,""));

}


public static void queenProblem()
{
    boolean[][] rooms=new boolean[4][4];
    int tnq = 4;

    // System.out.println(queensCombination2D(rooms,0,tnq,""));
    System.out.println(queensPermutation2D(rooms,tnq,""));
}

public static void Nqueen(){
    boolean[][] board=new boolean[10][10];
    int tnq=10;
    // System.out.println(Nqueen_01(board,0,tnq,""));
    // System.out.println(Nqueen_02(board,0,tnq,""));
    // System.out.println(Nqueen_03(board,0,tnq,""));
    // System.out.println(Nqueen_04(board,0,tnq,""));

    // int[][] board=new int[8][8];
    // for(int i=0;i<board.length;i++)
    //   Arrays.fill(board[i],-1);
    // System.out.println(nKnight(board,0,0,0));

    int n=10;
    ROW=new boolean[n];
    COL=new boolean[n];
    DIAG=new boolean[n+n-1];
    ADIAG=new boolean[n+n-1];
    
    long s= System.currentTimeMillis();
    System.out.println(Nqueen_05(n,n,0,n,""));
    long e= System.currentTimeMillis();
    System.out.println(e-s);


    // long s= System.currentTimeMillis();
    // System.out.println(Nqueen_06(n,n,0,n,""));
    // long e= System.currentTimeMillis();
    // System.out.println(e-s);

}

public static void solve(){
    // coinChange();
    // queenProblem();
    Nqueen();
}

}

File: 2019/LevelUp_1May20/lecture001_Recursion/l002.cpp
#include <iostream>
#include <vector>
using namespace std;

int coinChangePermutation_INF(vector<int> &arr, int lidx, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    for (int i = lidx; i < arr.size(); i++)
        if (tar - arr[lidx] >= 0)
            count += coinChangePermutation_INF(arr, 0, tar - arr[lidx], ans + to_string(arr[i]) + " ");

    return count;
}

int coinChangePermutation(vector<int> &arr, int lidx, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    for (int i = lidx; i < arr.size(); i++)
    {
        int temp = arr[i];
        if (arr[i] > 0 && tar - arr[i] >= 0)
        {
            arr[i] = -arr[i];
            count += coinChangePermutation(arr, 0, tar - temp, ans + to_string(temp) + " ");
            arr[i] = -arr[i];
        }
    }
    return count;
}

int coinChangeCombination_INF(vector<int> &arr, int lidx, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    for (int i = lidx; i < arr.size(); i++)
        if (tar - arr[i] >= 0)
            count += coinChangeCombination_INF(arr, i, tar - arr[i], ans + to_string(arr[i]) + " ");

    return count;
}

int coinChangeCombination(vector<int> &arr, int lidx, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    for (int i = lidx; i < arr.size(); i++)
        if (tar - arr[i] >= 0)
            count += coinChangeCombination(arr, i + 1, tar - arr[i], ans + to_string(arr[i]) + " ");

    return count;
}

//Queen's===========================================================

int queensCombination(vector<bool> &rooms, int room, int qpsf, int tnq, string ans) // qpsf: queen place so far.
{
    if (qpsf == tnq)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    for (int r = room; r < rooms.size(); r++)
        count += queensCombination(rooms, r + 1, qpsf + 1, tnq, ans + "Q" + to_string(qpsf) + "R" + to_string(r) + " ");

    return count;
}

int queensPermutation(vector<bool> &rooms, int room, int qpsf, int tnq, string ans) // qpsf: queen place so far.
{
    if (qpsf == tnq)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    for (int r = room; r < rooms.size(); r++)
        if (!rooms[r])
        {
            rooms[r] = true;
            count += queensPermutation(rooms, 0, qpsf + 1, tnq, ans + "Q" + to_string(qpsf) + "R" + to_string(r) + " ");
            rooms[r] = false;
        }

    return count;
}

//Sudoku.====================================================================

bool isSafeToPlaceNumber(vector<vector<char>> &board, int x, int y, int num)
{
    //row
    for (int i = 0; i < 9; i++)
        if (board[x][i] - '0' == num)
            return false;

    //col
    for (int i = 0; i < 9; i++)
        if (board[i][y] - '0' == num)
            return false;

    //mat
    x = (x / 3) * 3; // x-=x%3;
    y = (y / 3) * 3; // y-=y%3;
    for (int i = 0; i < 3; i++)
        for (int j = 0; j < 3; j++)
            if (board[x + i][y + j] - '0' == num)
                return false;

    return true;
}

vector<int> Srow;
vector<int> Scol;
vector<vector<int>> Smat;

bool solveSudoku_(vector<vector<char>> &board, vector<int> &calls, int idx)
{
    if (idx == calls.size())
        return true;

    int r = calls[idx] / 9;
    int c = calls[idx] % 9;
    for (int num = 1; num <= 9; num++)
    {
        int mask = 1 << num;
        if ((Srow[r] & mask) == 0 && (Scol[c] & mask) == 0 && (Smat[r / 3][c / 3] & mask) == 0)
        {
            board[r][c] = (char)(num + '0');
            Srow[r] ^= mask;
            Scol[c] ^= mask;
            Smat[r / 3][c / 3] ^= mask;

            if (solveSudoku_(board, calls, idx + 1))
                return true;

            board[r][c] = '.';
            Srow[r] ^= mask;
            Scol[c] ^= mask;
            Smat[r / 3][c / 3] ^= mask;
        }
    }
    return false;
}

//==============================================================================

void solveSudoku(vector<vector<char>> &board)
{
    vector<int> calls;
    Srow.assign(9, 0);
    Scol.assign(9, 0);
    Smat.assign(3, vector<int>(3, 0));

    for (int i = 0; i < 9; i++)
    {
        for (int j = 0; j < 9; j++)
        {
            if (board[i][j] == '.')
                calls.push_back(i * 9 + j);
            else
            {
                int num = board[i][j] - '0';
                int mask = 1 << num;
                Srow[i] ^= mask;
                Scol[j] ^= mask;
                Smat[i / 3][j / 3] ^= mask;
            }
        }
    }

    solveSudoku_(board, calls, 0);
}

//Crypto.==========================================================
string str1 = "send";
string str2 = "more";
string str3 = "money";
int isNumberUsed = 0;
vector<int> assignedNumber(26, -1);

int stringToNumber(string str)
{
    int res = 0;
    for (int i = 0; i < str.length(); i++)
        res = res * 10 + assignedNumber[str[i] - 'a'];
    return res;
}

int cryptoSolver_(string &str, int idx)
{
    if (idx == str.length())
    {
        int num1 = stringToNumber(str1);
        int num2 = stringToNumber(str2);
        int num3 = stringToNumber(str3);

        if (assignedNumber[str1[0] - 'a'] != 0 && assignedNumber[str2[0] - 'a'] != 0 && assignedNumber[str3[0] - 'a'] != 0 && num1 + num2 == num3)
        {
            cout << num1 << endl
                 << "+" << num2 << endl
                 << "------" << endl
                 << num3 << endl
                 << endl;
            return 1;
        }
        return 0;
    }

    int count = 0;
    for (int num = 0; num <= 9; num++)
    {
        int mask = 1 << num;
        if ((isNumberUsed & mask) == 0)
        {
            isNumberUsed ^= mask;
            assignedNumber[str[idx] - 'a'] = num;

            count += cryptoSolver_(str, idx + 1);

            isNumberUsed ^= mask;
            assignedNumber[str[idx] - 'a'] = -1;
        }
    }
    return count;
}

void crypto()
{
    string str = str1 + str2 + str3;
    vector<int> freq(26, 0);
    for (int i = 0; i < str.length(); i++)
        freq[str[i] - 'a']++;

    str = "";
    for (int i = 0; i < 26; i++)
        if (freq[i] > 0)
            str += (char)(i + 'a');

    // cout << str << endl;
    cout << cryptoSolver_(str, 0) << endl;
}

//SetProblem.===============================================================

int equiSet(vector<int> &arr, int idx, int set1, int set2, string set1S, string set2S)
{
    if (idx == arr.size())
    {
        if (set1 == set2)
        {
            cout << set1S + " = " + set2S << endl;
            return 1;
        }

        return 0;
    }

    int count = 0;
    count += equiSet(arr, idx + 1, set1 + arr[idx], set2, set1S + " " + to_string(arr[idx]), set2S);
    count += equiSet(arr, idx + 1, set1, set2 + arr[idx], set1S, set2S + " " + to_string(arr[idx]));

    return count;
}

int equiSet_02(vector<int> &arr, int idx, int set1, int set2, string set1S, string set2S)
{
    if (idx == arr.size())
    {
        if (set1 == set2 && set1 != 0)
        {
            cout << set1S + " = " + set2S << endl;
            return 1;
        }

        return 0;
    }
    int count = 0;
    count += equiSet_02(arr, idx + 1, set1 + arr[idx], set2, set1S + to_string(arr[idx]) + " ", set2S);
    count += equiSet_02(arr, idx + 1, set1, set2 + arr[idx], set1S, set2S + to_string(arr[idx]) + " ");

    count += equiSet_02(arr, idx + 1, set1, set2, set1S, set2S);

    return count;
}

//CROSS WORD.=================================================
vector<vector<char>> board{{'+', '-', '+', '+', '+', '+', '+', '+', '+', '+'},
                           {'+', '-', '+', '+', '+', '+', '+', '+', '+', '+'},
                           {'+', '-', '-', '-', '-', '-', '-', '-', '+', '+'},
                           {'+', '-', '+', '+', '+', '+', '+', '+', '+', '+'},
                           {'+', '-', '+', '+', '+', '+', '+', '+', '+', '+'},
                           {'+', '-', '-', '-', '-', '-', '-', '+', '+', '+'},
                           {'+', '-', '+', '+', '+', '-', '+', '+', '+', '+'},
                           {'+', '+', '+', '+', '+', '-', '+', '+', '+', '+'},
                           {'+', '+', '+', '+', '+', '-', '+', '+', '+', '+'},
                           {'+', '+', '+', '+', '+', '+', '+', '+', '+', '+'}};
vector<string> words = {"agra", "norway", "england", "gwalior"};

bool canPlaceHorizontal(string word, int x, int y)
{

    if (y == 0 && word.length() < board[0].size())
    {
        if (board[x][y + word.length()] != '+')
            return false;
    }
    else if ((y + word.length()) == board[0].size() && (word.length() != board[0].size()))
    {
        if (board[x][y - 1] != '+')
            return false;
    }
    else
    {
        if (((y - 1) >= 0 && board[x][y - 1] != '+') || ((y + word.length()) < board[0].size() && board[x][y + word.length()] != '+'))
            return false;
    }

    for (int i = 0; i < word.length(); i++)
    {
        if ((y + i) == board[0].size())
            return false;
        if (board[x][y + i] != '-' && board[x][y + i] != word[i])
            return false;
    }

    return true;
}

vector<bool> placeWordHorizontal(string word, int x, int y)
{
    vector<bool> loc(word.length(), false);
    for (int i = 0; i < word.length(); i++)
    {
        if (board[x][y + i] == '-')
        {
            loc[i] = true;
            board[x][y + i] = word[i];
        }
    }
    return loc;
}

void unPlaceWordHorizontal(string word, int x, int y, vector<bool> &loc)
{
    for (int i = 0; i < word.length(); i++)
        if (loc[i])
            board[x][y + i] = '-';
}

bool canPlaceVertical(string word, int x, int y)
{

    if (x == 0 && word.length() < board.size())
    {
        if (board[x + word.length()][y] != '+')
            return false;
    }
    else if ((x + word.length()) == board.size() && (word.length() != board.size()))
    {
        if (board[x - 1][y] != '+')
            return false;
    }
    else
    {
        if (((x - 1) >= 0 && board[x - 1][y] != '+') || ((x + word.length()) < board.size() && board[x + word.length()][y] != '+'))
            return false;
    }

    for (int i = 0; i < word.length(); i++)
    {
        if ((x + i) == board.size())
            return false;
        if (board[x + i][y] != '-' && board[x + i][y] != word[i])
            return false;
    }

    return true;
}

vector<bool> placeWordVertical(string word, int x, int y)
{
    vector<bool> loc(word.length(), false);
    for (int i = 0; i < word.length(); i++)
    {
        if (board[x + i][y] == '-')
        {
            loc[i] = true;
            board[x + i][y] = word[i];
        }
    }

    return loc;
}

void unPlaceWordVertical(string word, int x, int y, vector<bool> &loc)
{
    for (int i = 0; i < word.length(); i++)
        if (loc[i])
            board[x + i][y] = '-';
}

int crossWord_(int idx)
{
    if (idx == words.size())
    {
        for (int i = 0; i < board.size(); i++)
        {
            for (int j = 0; j < board[0].size(); j++)
                cout << board[i][j] << " ";
            cout << endl;
        }

        return 1;
    }

    string word = words[idx];
    int count = 0;
    for (int i = 0; i < board.size(); i++)
    {
        for (int j = 0; j < board[0].size(); j++)
        {
            if (board[i][j] == '-' || board[i][j] == word[0])
            {
                if (canPlaceHorizontal(word, i, j))
                {
                    vector<bool> loc = placeWordHorizontal(word, i, j);
                    count += crossWord_(idx + 1);
                    unPlaceWordHorizontal(word, i, j, loc);
                }

                if (canPlaceVertical(word, i, j))
                {
                    vector<bool> loc = placeWordVertical(word, i, j);
                    count += crossWord_(idx + 1);
                    unPlaceWordVertical(word, i, j, loc);
                }
            }
        }
    }
    return count;
}

void crossWord()
{
    crossWord_(0);
}

void SetProblem()
{
    vector<int> arr = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100};
    // cout << equiSet(arr, 1, 10, 0, "10 ", "") << endl;
    cout << equiSet_02(arr, 0, 0, 0, "", "") << endl;
}

void queenProblem()
{
    vector<bool> rooms(16, false);
    int tnq = 4;
    cout << queensCombination(rooms, 0, 0, tnq, "") << endl;
    // cout << queensPermutation(rooms, 0, 0, tnq, "") << endl;
}

void coinChange()
{
    // vector<int> arr{2, 3, 5, 7};
    // vector<int> arr{1,1,1,1,1};
    int tar = 10;
    // cout << coinChangePermutation_INF(arr,0, tar, "") << endl;
    // cout << coinChangePermutation(arr, 0, tar, "") << endl;

    // cout << coinChangeCombination_INF(arr,0, tar, "") << endl;
    // cout << coinChangeCombination(arr, 0, tar, "") << endl;
}

int main()
{
    // coinChange();
    // queenProblem();
    // crypto();
    // SetProblem();
    crossWord();
    return 0;
}
